### ISSUE TITLE
Fatal error: Unexpectedly found nil while implicitly unwrapping an Optional value

### ISSUE THREAD

**Issue by Surferdude667** [Daltron/NotificationBanner#307]
Hi there! This looks awesome, however I can't make it work.

I added this in my viewDidLoad()
<img width="950" alt="Skærmbillede 2020-06-24 kl  16 39 57" src="https://user-images.githubusercontent.com/15128640/85577471-8b056880-b639-11ea-8787-8f41244278be.png">

But when I run the app it crashes on this line:
<img width="1385" alt="Skærmbillede 2020-06-24 kl  16 40 13" src="https://user-images.githubusercontent.com/15128640/85577535-9c4e7500-b639-11ea-8e94-a18d4d128062.png">

**Comment by abeticlairon** [comment]
Same thing happens to me!

**Comment by burhanshakir** [comment]
Same here, but only seems to be happening on device, works fine on the simulator

**Comment by FakeAK** [comment]
Same thing here

**Comment by Surferdude667** [comment]
https://github.com/Daltron/NotificationBanner/issues/284#issuecomment-582237682

Look at this. I managed to fix it with this approach :) (However this is an issue)

**Comment by angi-betancourt** [comment]
I have the same issue but on a different line:
Fatal error: Unexpectedly found nil while unwrapping an Optional value: file /Users/isw/Documents/Home/AbarroTICs/frontend-ios/Código/Pods/NotificationBannerSwift/NotificationBanner/Classes/BaseNotificationBanner.swift, line 149
![Screen Shot 2020-07-30 at 13 55 58](https://user-images.githubusercontent.com/36274080/88962754-7f2a4900-d26c-11ea-8278-f2dc4e47db43.png)

It's an iphone app ios12+ with pod 'NotificationBannerSwift', '~> 3.0.0'
I have this class that centralize notification banner usage
![Screen Shot 2020-07-30 at 14 01 28](https://user-images.githubusercontent.com/36274080/88963400-666e6300-d26d-11ea-9006-137cdfab4580.png)
And I have this on an IBAction @ a ViewController
![Screen Shot 2020-07-30 at 14 02 00](https://user-images.githubusercontent.com/36274080/88963515-90278a00-d26d-11ea-8dd5-45322b7cbe12.png)

**Comment by belov1** [comment]
I have the same thing.
<img width="1028" alt="image" src="https://user-images.githubusercontent.com/30197503/90159639-61df9980-dd99-11ea-9911-b564f58763cd.png">

**Comment by jonathanfoster** [comment]
TL;DR: If you're having these strange issues, make sure you're instantiating and showing the banner on the same thread and avoid any shared objects without a locking mechanism.

I can into a similar issues (exact same as @Surferdude667 and @belov1) and it turned out to be a concurrency issues for me. I was using `NWPathMonitor` to monitor network connectivity and a `StatusBarNotificationBanner` to show status, both were class members. 

So the network monitor was being updated on a separate thread mean while the banner was updated on the main thread causing lots of strange behavior. Once I instantiated the the banner and called show on the main thread, everything was fine.

**Comment by misteu** [comment]
Hey there,
it does crash because the library depends on the view or at least the windon being existent. Unfortunately the library includes a lot of force unwrapping. 

For this particular issue, I found a solution. I will open a Pull Request for it soon.

**Comment by Neilfau** [comment]
Also had this issue, the library cannot find the window, I forked the project and made the following changes and it now works:

```
     private let appWindow: UIWindow? = {
         if #available(iOS 13.0, *) {
-            return UIApplication.shared.connectedScenes
-                .first { $0.activationState == .foregroundActive }
-                .map { $0 as? UIWindowScene }
-                .map { $0?.windows.first } ?? UIApplication.shared.delegate?.window ?? nil
+            return UIApplication.shared.windows.filter { $0.isKeyWindow }.first
         }
```

**Comment by findus** [comment]
For me the issue was that the app crashed if a modal dialog is being presented the moment the dialog gets displayed.
https://github.com/findus/NotificationBanner/commit/aefd75e8c916953f481d39afcdfede1487fae45c
This fixes the force unwrap of the view, but i am not sure if this screws up something else in the long run, if the subview does not get added, apart from that some notifications might get not  shown.

**Comment by DominikButz** [comment]
Same problem with UIAlert and in-app-purchase view controller. I tried to use the notification banner to display success or failure of an in-app-purchase - it crashes every time. I use it in a SwiftUI app, so it is scene-based. @findus solution worked for me.

**Comment by moh-abk** [comment]
> Same problem with UIAlert and in-app-purchase view controller. I tried to use the notification banner to display success or failure of an in-app-purchase - it crashes every time. I use it in a SwiftUI app, so it is scene-based. @Neilfau 's solution (in @findus fork) worked for me.

any chance you can share a snippet of how you use it with SwiftUI?

**Comment by DominikButz** [comment]
Hi, usage is not different from UIKit. 
you can see how this framework basically works in the show function:

    public func show(queuePosition: QueuePosition = .back,
                     bannerPosition: BannerPosition = .top,
                     queue: NotificationBannerQueue = NotificationBannerQueue.default,
                     on viewController: UIViewController? = nil) {
        parentViewController = viewController
        bannerQueue = queue
        show(placeOnQueue: true, queuePosition: queuePosition, bannerPosition: bannerPosition)
    }

So with SwiftUI you would typically let the UIViewController parameter nil (although you can try passing in a UIHostingController, e.g. if you started your app in iOS 13. 
if the view controller is nil the internal show function will place the banner into the app window, which should also be available in SwiftUI. 
The advantage of placing the banner in the app window is that the banner should always be on top - even if for example a SwiftUI sheet is currently showing. 
I hope this helps

**Comment by orchid-bloom** [comment]
I also reported an error on iOS 14.5
Fatal error: Unexpectedly found nil while implicitly unwrapping an Optional value: file NotificationBannerSwift/BaseNotificationBanner.swift, line 380

**Comment by alexookah** [comment]
@Daltron Should we make some PL to fix this?
